Libraries/LibWeb/Layout/BlockFormattingContext.cpp

```diff
@@ -496,30 +496,96 @@ void BlockFormattingContext::resolve_used_height_if_treated_as_auto(Box const& b
+        } else {
+            borders = computed_values.border_top().width + computed_values.border_bottom().width + computed_values.padding().top().to_px(box, box_state.containing_block_used_values()->content_height()) + computed_values.padding().bottom().to_px(box, box_state.containing_block_used_values()->content_height());
+        }
+        height = max(size - borders, height);
 
         // NOTE: The height of the root element when affected by this quirk is considered to be definite.
         box_state.set_has_definite_height(true);
     }
 
+    // https://quirks.spec.whatwg.org/#the-body-element-fills-the-html-element-quirk
+    // 3.7. The body element fills the html element quirk
+
+    // In quirks mode, if the document’s body element body is not null and if it matches the following conditions:
+    if (box.document().in_quirks_mode()
+        && box.dom_node()
+        && box.dom_node() == box.document().body()
+        // - The computed value of the width property of body is auto and body has a vertical writing mode,
+        && ((box.computed_values().width().is_auto() && box.text_flow_direction() == Gfx::Orientation::Vertical)
+            //   or the computed value of the height property of body is auto and body has a horizontal writing mode.
+            || (box.computed_values().height().is_auto() && box.text_flow_direction() == Gfx::Orientation::Horizontal))
+        // - The computed value of the position property of body is not absolute or fixed.
+        && !(box.computed_values().position() == CSS::Positioning::Absolute || box.computed_values().position() == CSS::Positioning::Fixed)
+        // - The computed value of the float property of body is none.
+        && box.computed_values().float_() == CSS::Float::None
+        // - body is not an inline-level element.
+        && !box.is_inline()
+        // - body is not a multi-column spanning element.
+        // FIXME: Implement this. Note: multiple columns seem to be required for this, so just reading the computed value of column-span is not enough.
+    ) {
+        // ...then body must have its border box size in the block flow direction set using the following algorithm:
+
+        // 1. Let margins be the sum of the used values of the margin-left and margin-right properties of body if body has a vertical writing mode, otherwise let margins be the sum of the
+        //    used values of the margin-top and margin-bottom properties of body.
+        CSSPixels margins;
+        if (box.text_flow_direction() == Gfx::Orientation::Vertical) {
+            margins = box_state.margin_left + box_state.margin_right;
+        } else {
+            margins = box_state.margin_top + box_state.margin_bottom;
+        }
+
+        // 2. Let size be the size of body’s parent element’s content box in the block flow direction minus margins.
+        auto size = box_state.containing_block_used_values()->content_height() - margins;
```

Comment:
Should we be using `content_width()` instead here if the block flow direction is horizontal (vertical text)?

If I'm reading the spec correctly, we should be updating the width rather than the height if the block flow direction is horizontal, which probably means that we need to implement this algorithm in a different place.

I think the same applies to the "html fills the viewport  quirk" above.